# 1. Extract data of Toronto neighborhoods from Wikipedia, clean and display the top 10 rows along with shape head¶


In [58]:
from bs4 import BeautifulSoup
import requests
import pandas as pd

In [78]:
# download url data from internet

wiki = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wikipedia_page = requests.get(wiki)

df_raw = pd.read_html(wikipedia_page.content, header=0)[0]
df_new = df_raw[df_raw.Borough != 'Not assigned']

df_new.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [81]:
#Find whether there is a "Not assigned" in Neighbourhood
df_new.loc[df_new.Neighborhood == 'Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

In [83]:
df_new.Neighborhood.replace('Not assigned',df_new.Borough,inplace=True)
df_new.head(8)

/opt/conda/envs/Python36/lib/python3.6/site-packages/pandas/core/generic.py:6586: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


Postal Code           Borough                                 Neighborhood
2          M3A        North York                                    Parkwoods
3          M4A        North York                             Victoria Village
4          M5A  Downtown Toronto                    Regent Park, Harbourfront
5          M6A        North York             Lawrence Manor, Lawrence Heights
6          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
8          M9A         Etobicoke      Islington Avenue, Humber Valley Village
9          M1B       Scarborough                               Malvern, Rouge
11         M3B        North York                                    Don Mills

In [88]:
# Group neighbours with same postcode
df_toronto = df_new.groupby(['Postal Code', 'Borough'])['Neighborhood'].apply(lambda x: ', '.join(x))
df_toronto = df_toronto.reset_index()
df_toronto.rename(columns = {'Postcode':'Postal Code'}, inplace = True)
df_toronto.rename(columns = {'Neighbourhood':'Neighborhood'}, inplace = True)
df_toronto.head()

Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

In [89]:
df_toronto.shape

(103, 3)

# 2. Latitude and Longitude

In [90]:
url = 'http://cocl.us/Geospatial_data'
df_geo=pd.read_csv(url)
df_geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [91]:

#check the sshape o the csv file
df_geo.shape

(103, 3)

In [93]:
#Both tables have the same chape. Now we can join new colums to our data.
df_toronto = df_toronto.join(df_geo.set_index('Postal Code'), on='Postal Code')
df_toronto.head

<bound method NDFrame.head of     Postal Code           Borough  \
0           M1B       Scarborough   
1           M1C       Scarborough   
2           M1E       Scarborough   
3           M1G       Scarborough   
4           M1H       Scarborough   
5           M1J       Scarborough   
6           M1K       Scarborough   
7           M1L       Scarborough   
8           M1M       Scarborough   
9           M1N       Scarborough   
10          M1P       Scarborough   
11          M1R       Scarborough   
12          M1S       Scarborough   
13          M1T       Scarborough   
14          M1V       Scarborough   
15          M1W       Scarborough   
16          M1X       Scarborough   
17          M2H        North York   
18          M2J        North York   
19          M2K        North York   
20          M2L        North York   
21          M2M        North York   
22          M2N        North York   
23          M2P        North York   
24          M2R        North York   
25      

The dataframe has 103 Postal codes but it has 212 rows, because each Postal code can present more than one neighborhood (210 in total). Therefore, the dataframe should be group by the Postal code, ending with a dataframe with 103 rows.

# Use the foursquere API to segment and cluster the neighborhoods of Toronto

In [94]:

!conda install -c conda-forge geocoder --yes
import geocoder
from geopy.geocoders import Nominatim 

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geocoder


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    geocoder-1.38.1            |             py_1          53 KB  conda-forge
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    ratelim-0.1.6              |             py_2           6 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geocoder:        1.38.1-py_1       conda-forge
    python_abi:    

In [95]:
# print map with all broughts
import folium

# create map of Toronto using latitude and longitude values
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(map_Toronto)  
    
map_Toronto

In [97]:
# define foursquare credentials and versions
CLIENT_ID = 'QYBLOT2MSO3Q1LHAPJUNP4UP45CJ15HBLEIKNEVQFRSLCDGX' # your Foursquare ID
CLIENT_SECRET = 'HLBTJXCN5LEZPCWLNEF1MZGQ5UBCQ34P4DTMQ4YSB4BDGYYX' # your Foursquare Secret
VERSION = '20180604' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: QYBLOT2MSO3Q1LHAPJUNP4UP45CJ15HBLEIKNEVQFRSLCDGX
CLIENT_SECRET:HLBTJXCN5LEZPCWLNEF1MZGQ5UBCQ34P4DTMQ4YSB4BDGYYX


In [98]:
# Explore the data in 500 meters range from our first entry

neighborhood_latitude = df_toronto.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df_toronto.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df_toronto.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Malvern, Rouge are 43.806686299999996, -79.19435340000001.


In [99]:
# Create Get URL
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=QYBLOT2MSO3Q1LHAPJUNP4UP45CJ15HBLEIKNEVQFRSLCDGX&client_secret=HLBTJXCN5LEZPCWLNEF1MZGQ5UBCQ34P4DTMQ4YSB4BDGYYX&v=20180604&ll=43.806686299999996,-79.19435340000001&radius=500&limit=100'

In [100]:

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5eef395bb9a389001b8777c9'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Malvern',
  'headerFullLocation': 'Malvern, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 1,
  'suggestedBounds': {'ne': {'lat': 43.8111863045, 'lng': -79.18812958073042},
   'sw': {'lat': 43.80218629549999, 'lng': -79.2005772192696}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bb6b9446edc76b0d771311c',
       'name': 'Wendy’s',
       'location': {'crossStreet': 'Morningside & Sheppard',
        'lat': 43.80744841934756,
        'lng': -79.19905558052072,
        'labeledLatLngs': [{'label': 'display'

In [101]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

We can see that we had only 1 response for our first entry

In [102]:
import json
from pandas.io.json import json_normalize

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat        lng
0  Wendy’s  Fast Food Restaurant  43.807448 -79.199056

In [106]:
# Gernalize to obtain the venues from all neighborshoods in TORONTO
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [134]:
toronto_venues = getNearbyVenues(names=df_toronto['Neighborhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Parkwoods
Don Mills
Don Mills
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto, Broadview North (Old East York)
The Danforth West, 

In [135]:
print(toronto_venues.shape)
toronto_venues.head()

(2132, 7)


Neighbourhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1  Rouge Hill, Port Union, Highland Creek              43.784535   
2  Rouge Hill, Port Union, Highland Creek              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                        Venue  Venue Latitude  \
0              -79.194353                      Wendy’s       43.807448   
1              -79.160497  Great Shine Window Cleaning       43.783145   
2              -79.160497        Royal Canadian Legion       43.782533   
3              -79.188711               RBC Royal Bank       43.766790   
4              -79.188711            G & G Electronics       43.765309   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.157431          Home Service  
2       -79.163085                   Bar  
3       -79.191151                  Bank  
4       -79.191537     Electronics Store

In [136]:
#How many neighbours value
toronto_venues.groupby('Neighbourhood').count()

Neighborhood Latitude  \
Neighbourhood                                                               
Agincourt                                                               4   
Alderwood, Long Branch                                                  7   
Bathurst Manor, Wilson Heights, Downsview North                        23   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      23   
Berczy Park                                                            58   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           22   
Business reply mail Processing Centre, South Ce...                     18   
CN Tower, King and Spadina, Railway Lands, Harb...                     14   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  13   
Cedarbrae                                                               8   
Central Bay Street                                                     65   
Christie                                                               17   
Church and Wellesley                                                   76   
Clarks Corners, Tam O'Shanter, Sullivan                                14   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        8   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        5   
Don Mills                                                              26   
Dorset Park, Wexford Heights, Scarborough Town ...                      6   
Downsview                                                              16   
Dufferin, Dovercourt Village                                           17   
East Toronto, Broadview North (Old East York)                           4   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      8   
Fairview, Henry Farm, Oriole                                           65   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         6   
Garden District, Ryerson                                              100   
Glencairn                                                               6   
Golden Mile, Clairlea, Oakridge                                         8   
Guildwood, Morningside, West Hill                                       8   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
Hillcrest Village                                                       5   
Humber Summit                                                           1   
Humberlea, Emery                                                        3   
Humewood-Cedarvale                                                      4   
India Bazaar, The Beaches West                                         20   
Kennedy Park, Ionview, East Birchmount Park                             5   
Kensington Market, Chinatown, Grange Park                              63   
Kingsview Village, St. Phillips, Martin Grove G...                      4   
Lawrence Manor, Lawrence Heights                                       11   
Lawrence Park                                                           3   
Leaside                                                                32   
Little Portugal, Trinity                                               44   
Malvern, Rouge                                                          1   
Milliken, Agincourt North, Steeles East, L'Amor.

How many categories can we find?

In [137]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  \
0                          Malvern, Rouge                  0   
1  Rouge Hill, Port Union, Highland Creek                  0   
2  Rouge Hill, Port Union, Highland Creek                  0   
3       Guildwood, Morningside, West Hill                  0   
4       Guildwood, Morningside, West Hill                  0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0                 0                   0            0              0   
1                 0                   0            0              0   
2                 0                   0            0              0   
3                 0                   0            0              0   
4                 0                   0            0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    1               0   
3          0           0           0       0     1    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0                 0                 0                   0      0   
1                 0                 0                   0      0   
2                 0                 0                   0      0   
3                 0                 0                   0      0   
4                 0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0       0           0              0          0         0   
1       0           0              0          0         0   
2       0           0              0          0         0   
3       0           0              0          0         0   
4       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0   

In [138]:
toronto_onehot.shape

(2132, 271)

In [139]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.000000   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.000000   
4             0.0              0.0               0.0             0.043478   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0          0.0         0.0                  0.0   
1           0.0       0.0          0.0         0.0                  0.0   
2           0.0       0.0          0.0         0.0                  0.0   
3           0.0       0.0          0.0         0.0                  0.0   
4           0.0       0.0          0.0         0.0                  0.0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0               0.0                 0.0          0.0            0.0   
1               0.0                 0.0          0.0            0.0   
2               0.0                 0.0          0.0            0.0   
3               0.0                 0.0          0.0            0.0   
4               0.0                 0.0          0.0            0.0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery      Bank  Bar  Baseball Field  \
0        0.0         0.0         0.0     0.0  0.000000  0.0             0.0   
1        0.0         0.0         0.0     0.0  0.000000  0.0             0.0   
2        0.0         0.0         0.0     0.0  0.086957  0.0             0.0   
3        0.0         0.0         0.0     0.0  0.250000  0.0             0.0   
4        0.0         0.0         0.0     0.0  0.000000  0.0             0.0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0               0.0               0.0                 0.0    0.0   
1               0.0               0.0                 0.0    0.0   
2               0.0               0.0                 0.0    0.0   
3               0.0               0.0                 0.0    0.0   
4               0.0               0.0                 0.0    0.0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0              0.0       0.0         0.0                 0.0        0.0   
1              0.0       0.0         0.0                 0.0        0.0   
2              0.0       0.0         0.0                 0.0        0.0   
3              0.0       0.0         0.0                 0.0        0.0   
4              0.0       0.0         0.0                 0.0        0.0   

   Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0     0.0         0.0            0.0        0.0       0.0   
1     0.0         0.0            0.0        0.0       0.0   
2     0.0         0.0            0.0        0.0       0.0   
3     0.0         0.0            0.0        0.0       0.0   
4     0.0         0.0            0.0        0.0       0.0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                   0.0            0.25      0.0     0.000000   
1                   0.

In [140]:
toronto_grouped.shape

(95, 271)

In [141]:
# Get the top 10 for each row
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [142]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant          Skating Rink        Breakfast Spot   
1                Coffee Shop        Sandwich Place              Pharmacy   
2                Coffee Shop             Gift Shop                 Diner   
3                       Bank    Chinese Restaurant   Japanese Restaurant   
4                Coffee Shop            Restaurant    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   
1                   Pub                   Gym               Airport Lounge   
2        Ice Cream Shop           Supermarket             Sushi Restaurant   
3      Department Store    Dim Sum Restaurant                        Diner   
4             Juice Bar                   Pub                         Café   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Dumpling Restaurant          Dessert Shop              Drugstore  
1                Fabric Shop  Ethiopian Restaurant      Electronics Store  
2  Middle Eastern Restaurant            Restaurant      Mobile Phone Shop  
3             Discount Store   Distribution Center                Dog Run  
4           Sushi Restaurant     Indian Restaurant            Pizza Place

### Custer Neighborhoods

In [146]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2], dtype=int32)

### Merge the dataframe with the top 10 and the cluster for each neighbourhood

In [148]:

# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             1.0  Fast Food Restaurant           Yoga Studio   
1 -79.160497             2.0                   Bar          Home Service   
2 -79.188711             2.0   Rental Car Location         Moving Target   
3 -79.216917             2.0           Coffee Shop     Korean Restaurant   
4 -79.239476             2.0                Bakery      Hakka Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0            Donut Shop    Dim Sum Restaurant                 Diner   
1           Yoga Studio      Department Store    Dim Sum Restaurant   
2        Breakfast Spot        Medical Center          Intersection   
3           Yoga Studio      Doner Restaurant    Dim Sum Restaurant   
4                  Bank    Athletics & Sports  Caribbean Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Discount Store   Distribution Center               Dog Run   
1                 Diner        Discount Store   Distribution Center   
2                  Bank    Mexican Restaurant     Electronics Store   
3                 Diner        Discount Store   Distribution Center   
4       Thai Restaurant           Gas Station   Fried Chicken Joint   

  9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant              Drugstore  
1               Dog Run       Doner Restaurant  
2        Cosmetics Shop        Coworking Space  
3               Dog Run              Drugstore  
4        Discount Store     Dim Sum Restaurant

In [149]:
toronto_merged[toronto_merged['Cluster Labels'].isnull()]

Postal Code      Borough                             Neighborhood  \
16         M1X  Scarborough                              Upper Rouge   
20         M2L   North York                 York Mills, Silver Hills   
21         M2M   North York                  Willowdale, Newtonbrook   
93         M9A    Etobicoke  Islington Avenue, Humber Valley Village   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
16  43.836125 -79.205636             NaN                   NaN   
20  43.757490 -79.374714             NaN                   NaN   
21  43.789053 -79.408493             NaN                   NaN   
93  43.667856 -79.532242             NaN                   NaN   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
16                   NaN                   NaN                   NaN   
20                   NaN                   NaN                   NaN   
21                   NaN                   NaN                   NaN   
93                   NaN                   NaN                   NaN   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
16                   NaN                   NaN                   NaN   
20                   NaN                   NaN                   NaN   
21                   NaN                   NaN                   NaN   
93                   NaN                   NaN                   NaN   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
16                   NaN                   NaN                    NaN  
20                   NaN                   NaN                    NaN  
21                   NaN                   NaN                    NaN  
93                   NaN                   NaN                    NaN

Plot the Clusters

In [150]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighborhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examin Clusters

### Cluster 1

In [151]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14       Scarborough             0.0                  Park   
50  Downtown Toronto             0.0                  Park   
74              York             0.0                  Park   
90         Etobicoke             0.0                 River   
98              York             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14            Playground           Yoga Studio               Dog Run   
50            Playground                 Trail           Yoga Studio   
74         Women's Store                  Pool      Department Store   
90                  Park               Dog Run      Department Store   
98         Deli / Bodega  Ethiopian Restaurant     Electronics Store   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14             Department Store          Dessert Shop    Dim Sum Restaurant   
50                      Dog Run      Department Store          Dessert Shop   
74                 Dessert Shop    Dim Sum Restaurant                 Diner   
90                 Dessert Shop    Dim Sum Restaurant                 Diner   
98  Eastern European Restaurant   Dumpling Restaurant             Drugstore   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14                 Diner        Discount Store    Distribution Center  
50    Dim Sum Restaurant                 Diner         Discount Store  
74        Discount Store   Distribution Center                Dog Run  
90        Discount Store   Distribution Center            Yoga Studio  
98            Donut Shop      Doner Restaurant                Dog Run

### Cluster 2

In [152]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  Scarborough             1.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0            Donut Shop    Dim Sum Restaurant                 Diner   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Discount Store   Distribution Center               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant              Drugstore

### Cluster 3

In [153]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels          1st Most Common Venue  \
1         Scarborough             2.0                            Bar   
2         Scarborough             2.0            Rental Car Location   
3         Scarborough             2.0                    Coffee Shop   
4         Scarborough             2.0                         Bakery   
5         Scarborough             2.0                  Grocery Store   
6         Scarborough             2.0                  Train Station   
7         Scarborough             2.0                         Bakery   
8         Scarborough             2.0            American Restaurant   
9         Scarborough             2.0                College Stadium   
10        Scarborough             2.0              Indian Restaurant   
11        Scarborough             2.0      Middle Eastern Restaurant   
12        Scarborough             2.0                         Lounge   
13        Scarborough             2.0                       Pharmacy   
15        Scarborough             2.0             Chinese Restaurant   
17         North York             2.0       Mediterranean Restaurant   
18         North York             2.0                 Clothing Store   
19         North York             2.0                           Café   
22         North York             2.0               Ramen Restaurant   
23         North York             2.0                           Park   
24         North York             2.0                  Grocery Store   
25         North York             2.0                           Park   
26         North York             2.0                    Coffee Shop   
27         North York             2.0                    Coffee Shop   
28         North York             2.0                           Bank   
29         North York             2.0             Miscellaneous Shop   
30         North York             2.0               Business Service   
31         North York             2.0               Business Service   
32         North York             2.0               Business Service   
33         North York             2.0               Business Service   
34         North York             2.0                   Intersection   
35          East York             2.0                    Pizza Place   
36          East York             2.0                   Skating Rink   
37       East Toronto             2.0                            Pub   
38          East York             2.0                    Coffee Shop   
39          East York             2.0              Indian Restaurant   
40          East York             2.0              Convenience Store   
41       East Toronto             2.0               Greek Restaurant   
42       East Toronto             2.0                 Sandwich Place   
43       East Toronto             2.0                           Café   
44    Central Toronto             2.0                           Park   
45    Central Toronto             2.0                            Gym   
46    Central Toronto             2.0                 Clothing Store   
47    Central Toronto             2.0                    Pizza Place   
48    Central Toronto             2.0                   Tennis Court   
49    Central Toronto             2.0                    Coffee Shop   
51   Downtown Toronto             2.0                    Coffee Shop   
52   Downtown Toronto             2.0                    Coffee Shop   
53   Downtown Toronto             2.0                    Coffee Shop   
54   Downtown Toronto             2.0                 Clothing Store   
55   Downtown Toronto             2.0                    Coffee Shop   
56   Downtown Toronto             2.0                    Coffee Shop   
57   Downtown Toronto             2.0                    Coffee Shop   
58   Downtown Toronto             2.0                    Coffee Shop   
59   Downtown Toronto             2.0                    Coffee Shop   
60   Downtown Toronto             2.0                 

### Cluster 4

In [154]:

toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
94  Etobicoke             3.0           Golf Course           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
94            Donut Shop    Dim Sum Restaurant                 Diner   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
94        Discount Store   Distribution Center               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
94      Doner Restaurant              Drugstore

### Cluster 5

In [155]:

toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
81        York             4.0     Convenience Store           Pizza Place   
96  North York             4.0           Pizza Place           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
81      Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
96      Doner Restaurant          Dessert Shop    Dim Sum Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
81                 Diner        Discount Store   Distribution Center   
96                 Diner        Discount Store   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
81               Dog Run            Yoga Studio  
96               Dog Run             Donut Shop